vita3k/renderer/src/vulkan/texture.cpp

```diff
@@ -0,0 +1,351 @@
+        // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSamplerCreateInfo.html
+        // if there is no mipmap, set maxLod to 0.25 so it uses both the magnification or minification filter when needed
+        .maxLod = mipmap_enabled ? static_cast<float>(mip_count) : 0.25f,
+        .unnormalizedCoordinates = VK_FALSE,
+    };
+    return state.device.createSampler(sampler_info);
+}
+
+// add an alpha channel to u8u8u8 textures
+static void *add_alpha_channel(const void *pixels, const uint32_t width, const uint32_t height, std::vector<uint8_t> &data) {
+    data.resize(width * height * 4);
+
+    const uint8_t *src = reinterpret_cast<const uint8_t *>(pixels);
+    uint8_t *dst = data.data();
+    for (int y = 0; y < height; y++) {
+        for (int x = 0; x < width; x++) {
+            dst[0] = src[0];
+            dst[1] = src[1];
+            dst[2] = src[2];
+            // set 1.0 as the alpha channel
+            dst[3] = 255;
+
+            src += 3;
+            dst += 4;
+        }
+    }
+
+    return data.data();
+}
+
+void upload_bound_texture(VKTextureCacheState &cache, SceGxmTextureBaseFormat base_format, uint32_t width, uint32_t height,
+    uint32_t mip_index, const void *pixels, int face, bool is_compressed, size_t pixels_per_stride) {
+    vkutil::Image &image = *cache.current_texture;
+
+    const void *text_data = pixels;
+    std::vector<uint8_t> temp_data;
+    if (base_format == SCE_GXM_TEXTURE_BASE_FORMAT_U8U8U8) {
+        text_data = add_alpha_channel(pixels, width, height, temp_data);
+        base_format = SCE_GXM_TEXTURE_BASE_FORMAT_U8U8U8U8;
+    }
+
+    if (face > 0)
+        face--;
+
+    if (pixels_per_stride == 0)
+        pixels_per_stride = width;
+
+    vk::DeviceSize upload_size;
+    if (is_compressed) {
+        upload_size = renderer::texture::get_compressed_size(base_format, width, height);
+    } else {
+        size_t bpp = renderer::texture::bits_per_pixel(base_format);
+        size_t bytes_per_pixel = (bpp + 7) >> 3;
+        upload_size = pixels_per_stride * height * bytes_per_pixel;
+    }
+
+    if (cache.alloc_info.size < upload_size) {
+        // we need to create a bigger buffer
+        if (cache.staging_buffer) {
+            // delete the previous one if there is
+            cache.state.allocator.destroyBuffer(cache.staging_buffer, cache.alloc);
+        }
+
+        vk::BufferCreateInfo buffer_info{
+            .size = upload_size,
+            .usage = vk::BufferUsageFlagBits::eTransferSrc,
+            .sharingMode = vk::SharingMode::eExclusive
+        };
+        std::tie(cache.staging_buffer, cache.alloc) = cache.state.allocator.createBuffer(buffer_info, vkutil::vma_mapped_alloc, cache.alloc_info);
+    }
+
+    memcpy(cache.alloc_info.pMappedData, text_data, upload_size);
+    vk::CommandBuffer cmd_buffer = vkutil::create_single_time_command(cache.state.device, cache.state.transfer_command_pool);
```

Comment:
This is totally fine, but we can do some fancy optimization later by using tansfer queue -- we might have to batch another stream of commands with semaphore syncing between render command list.

Reply:
Yes, the code for uploading texture is really poorly optimized (`create_single_time_command` should not be called for anything that happens more than once a frame). I plan on uploading textures on a parallel thread using a transfer queue, but I will do this in a separate PR after this is merged.